worlds/kh2/Rules.py

```diff
@@ -1,96 +1,1839 @@
-        if world.FinalXemnas[player]:
-            world.completion_condition[player] = lambda state: state.kh_victory(player)
+            add_rule(
+                    self.world.multiworld.get_location(LocationName.FinalXemnas, self.player),
+                    lambda state: state.has(ItemName.Bounty, self.player, self.world.multiworld.BountyRequired[self.player].value) and
+                                  state.has(ItemName.LuckyEmblem, self.player, self.world.multiworld.LuckyEmblemsRequired[self.player].value))
+
+            if self.world.multiworld.FinalXemnas[self.player]:
+                self.world.multiworld.completion_condition[self.player] = lambda state: state.has(ItemName.Victory, self.player, 1)
+            else:
+                self.world.multiworld.completion_condition[self.player] = lambda state: state.has(ItemName.Bounty, self.player, self.world.multiworld.BountyRequired[self.player].value) and \
+                                                                                        state.has(ItemName.LuckyEmblem, self.player, self.world.multiworld.LuckyEmblemsRequired[self.player].value)
+
+    def kh2_set_count_sum(self, item_name_set: set, state: CollectionState) -> int:
+        """
+        Returns the sum of state.count() for each item in the set.
+        """
+        return sum(
+                [state.count(item_name, self.player) for item_name in list(item_name_set) if
+                 state.count(item_name, self.player)]
+        )
+
+    def kh2_list_any_sum(self, list_of_item_name_list: list, state: CollectionState) -> int:
+        """
+        Returns sum that increments by 1 if state.has_any
+        """
+        return sum(
+                [1 for item_list in list_of_item_name_list if
+                 state.has_any(set(item_list), self.player)]
+        )
+
+    def kh2_dict_count(self, item_name_to_count: dict, state: CollectionState) -> bool:
+        """
+        simplifies count to a dictionary.
+        """
+        return all(
+                [state.count(item_name, self.player) >= item_amount for item_name, item_amount in
+                 item_name_to_count.items() if item_name]
+        )
+
+    def kh2_dict_one_count(self, item_name_to_count: dict, state: CollectionState) -> int:
+        """
+        simplifies count to a dictionary.
+        """
+        return sum(
+                [1 for item_name, item_amount in
+                 item_name_to_count.items() if state.count(item_name, self.player) >= item_amount]
+        )
+
+    def kh2_can_reach_any(self, loc_set: list, state: CollectionState):
+        """
+        Can reach any locations in the set.
+        """
+        return any(
+                [self.kh2_can_reach(location, state) for location in
+                 loc_set]
+        )
+
+    def kh2_can_reach_all(self, loc_list: list, state: CollectionState):
+        """
+        Can reach all locations in the set.
+        """
+        return all(
```

Comment:
i think it'd be reasonable to `self.kh2_can_reach(loc, state) for loc in loc_list`